src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -629,6 +633,119 @@ theorem shiftLeft_shiftLeft {w : Nat} (x : BitVec w) (n m : Nat) :
+      omega
+
+/-- The MSB of a bitvector is `false` iff its value is smaller than 2^(half the bitwidth).
+  This is the contrapositive of `BitVec.msb_eq_true_iff_large`.
+  TODO: this can be massively cleaned up via `contrapose`, but unfortunately this is in mathli  -/
+theorem msb_eq_false_iff_small {x : BitVec w} :
+    x.msb = false ↔ 2 * x.toNat < 2^w  := by
+  constructor
+  · intros h
+    by_cases hx : (2 * x.toNat ≥ 2^w) <;> try omega
+    exfalso
+    simp_all [msb_eq_true_iff_large (x := x).mpr (by omega)]
+  · intros h
+    rcases hmsb : x.msb with rfl | rfl <;> try rfl
+    exfalso
+    have hcontra := msb_eq_true_iff_large (x := x).mp hmsb
+    omega
+
+/-- Equation theorem for 'Int.sub' when both arguments are 'negSucc' -/
+theorem Int.toNat_sub_toNat_eq_negSucc_ofLt {n m : Nat} (hlt : n < m) :
+  (n : Int) - (m : Int) = (Int.negSucc (m - 1 - n)) := by
+  rw [Int.negSucc_eq] -- TODO: consider adding this to omega cleanup set.
+  omega
+
+/-- Equation theorem for 'Int.mod' -/
+private theorem Int.mod_negSucc_eq (m : Nat) (n : Int) :
+  (Int.negSucc m) % n = Int.subNatNat (Int.natAbs n) (Nat.succ (m % Int.natAbs n)) := rfl
+
+/-- if the msb is false, the arithmetic shift right equals logical shift right -/
+theorem sshiftRight_eq_of_msb_false {x : BitVec w} {s : Nat} (h : x.msb = false) :
+    (x.sshiftRight s) = x >>> s := by
+  rcases w with rfl | w
+  · apply Subsingleton.elim
+  · apply BitVec.eq_of_toNat_eq
+    rw [BitVec.sshiftRight_eq]
+    rw [BitVec.toInt_eq_toNat_cond]
+    have hxbound : 2 * x.toNat < 2 ^ (w + 1) := BitVec.msb_eq_false_iff_small.mp h
+    simp only [hxbound, ↓reduceIte, Int.shiftRight_ofNat, Int.ofNat_eq_coe, ofInt_natCast,
+      toNat_ofNat, toNat_ushiftRight]
+    rw [Nat.mod_eq_of_lt]
+    rw [Nat.shiftRight_eq_div_pow]
+    suffices x.toNat / 2^s ≤ x.toNat by
+      apply Nat.lt_of_le_of_lt this x.isLt
+    apply Nat.div_le_self
+
+/-- If the msb is `true`, the arithmetic shift right equals negating, then logical shifting right, then negating again.
+  The double negation preserves the lower bits that have been shifted,
+  and the outer negation ensures that the high bits are '1'. -/
+theorem sshiftRight_eq_of_msb_true {x : BitVec w} {s : Nat} (h : x.msb = true) :
+    (x.sshiftRight s) = ~~~((~~~x) >>> s) := by
+  rcases w with rfl | w
+  · apply BitVec.eq_of_toNat_eq
+    simp
+  · apply BitVec.eq_of_toNat_eq
+    rw [BitVec.sshiftRight_eq]
+    rw [BitVec.toInt_eq_toNat_cond]
+    have hxbound : (2 * x.toNat ≥ 2 ^ (w + 1)) := BitVec.msb_eq_true_iff_large.mp h
+    have hxbound' : ¬ (2 * x.toNat < 2 ^ (w + 1)) := by omega
+    simp only [hxbound', ↓reduceIte, toNat_ofInt, toNat_not, toNat_ushiftRight]
+    rw [Int.toNat_sub_toNat_eq_negSucc_ofLt (by omega)]
+    rw [Int.shiftRight_negSucc]
+    rw [Int.mod_negSucc_eq]
+    simp only [Int.natAbs_ofNat, Nat.succ_eq_add_one]
+    rw [Int.subNatNat_of_le]
+    simp only [Int.toNat_ofNat]
+    rw [Nat.mod_eq_of_lt]
+    omega
```

Comment:
```suggestion
    rw [Nat.mod_eq_of_lt (by omega)]
```
Currenlty, it's not super clear `omega` is closing a subgoal

Reply:
@alexkeizer Note that the proof that `Nat.mod_eq_of_lt` is correct is the *larger* piece below:

```lean
    · rw [Nat.shiftRight_eq_div_pow]
      suffices (2 ^ (w + 1) - 1 - x.toNat) / 2 ^ s < 2 ^ (w + 1) - 1 by omega
      apply Nat.lt_of_le_of_lt (Nat.div_le_self _ _) (by omega)
```

--- 

I tried to clean the implementation by using a more structured proof style

```lean
    rw [Nat.mod_eq_of_lt]
    · omega
    · rw [Nat.shiftRight_eq_div_pow]
      suffices (2 ^ (w + 1) - 1 - x.toNat) / 2 ^ s < 2 ^ (w + 1) - 1 by omega
      apply Nat.lt_of_le_of_lt (Nat.div_le_self _ _) (by omega)
```